processing/src/main/java/io/druid/query/dimension/ListFilteredDimensionSpec.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.dimension;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import com.metamx.common.StringUtils;
+import io.druid.query.filter.DimFilterCacheHelper;
+import io.druid.segment.DimensionSelector;
+import io.druid.segment.data.IndexedInts;
+import io.druid.segment.data.ListBasedIndexedInts;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ */
+public class ListFilteredDimensionSpec extends BaseFilteredDimensionSpec
+{
+
+  private static final byte CACHE_TYPE_ID = 0x3;
+
+  private final List<String> values;
+
+  private volatile Set<Integer> matched;
+
+  public ListFilteredDimensionSpec(
+      @JsonProperty("delegate") DimensionSpec delegate,
+      @JsonProperty("values") List<String> values
+  )
+  {
+    super(delegate);
+
+    Preconditions.checkArgument(values != null && values.size() > 0, "values list must be non-empty");
+    this.values = values;
+  }
+
+  @JsonProperty
+  public List<String> getValues()
+  {
+    return values;
+  }
+
+  @Override
+  public DimensionSelector decorate(final DimensionSelector selector)
+  {
+    if (selector == null) {
+      return selector;
+    }
+
+    if (matched == null) {
+      matched = new HashSet<>(values.size());
```

Comment:
If this is applied to multiple segments at the same time, the `HashSet` reference is gonna be changing out from under us.  Would be better to make a new one here and then close over the `DimensionSelector`